### PR TITLE
Make DataTable's groupBy togglable

### PR DESCRIPTION
**Pull request by watiko**
#### What does this PR do?

Make DataTable's groupBy togglable.

#### Where should the reviewer start?

`DataTable.js`

#### What testing has been done on this PR?

Check if the error below was not thrown when toggle `groupBy` props between `undefined` and `"a"`.

>Uncaught [TypeError: Cannot read property 'expanded' of undefined]

### CODE REVIEW

File: src/js/components/DataTable/Body.js [hunk 1]
```diff
@@ -204,7 +204,7 @@ const Body = forwardRef(
               const isRowExpanded = rowExpand && rowExpand.includes(index);
               return (
                 <Row
-                  key={primaryValue || index}
```
Comment (watiko) [inline src/js/components/DataTable/Body.js:207]: `primaryValue` is not always unique depending on what is passed.

**Comment by ShimiSun** [comment]
Hi buddy, thanks for this PR!
We are missing some context about the original issue and what this PR is trying to solve, can you please elaborate?

**Comment by watiko** [comment]
@ShimiSun Here is the context.

I has an app with a togglable table component using (old) Grommet's DataTable and the table component can be toggle between plain state (`groupBy={undefined}`) and grouped by state(`groupBy="some column name"`).
When I was trying to update Grommet's version, I had a trouble about the application's crash by toggling table state.

>Uncaught [TypeError: Cannot read property 'expanded' of undefined]

After some investigations, I found that the root cause of the problem. `groupState` can be inconsistent with `groups` and `groupBy`.

**Comment by ShimiSun** [comment]
@watiko Can you please share a reproducible code of the issue you are describing? It will help us understand your fix better. Here is a [template](https://codesandbox.io/s/m7mml8l0zj)

**Comment by watiko** [comment]
@ShimiSun Here is [the reproducible code](https://codesandbox.io/s/grommet-v2-template-forked-njh96?file=/index.js)

```javascript
function CrashOnToggling() {
  const [groupBy, setGroupBy] = React.useState();
  const toggle = () => setGroupBy(groupBy === undefined ? "a" : undefined);

  return (
    <Grommet>
      <button type="button" onClick={toggle}>
        toggle
      </button>
      <DataTable
        columns={[
          { property: "a", header: "A" },
          { property: "b", header: "B" }
        ]}
        data={[
          { a: "one", b: 1.1 },
          { a: "one", b: 1.2 },
          { a: "two", b: 2.1 },
          { a: "two", b: 2.2 }
        ]}
        groupBy={groupBy}
      />
    </Grommet>
  );
}
```

**Comment by Naguumo** [comment]
I have been having this same issue with my use case and would love to see more attention from the team on this issue.

File: src/js/components/DataTable/DataTable.js [hunk 2]
```diff
@@ -53,6 +53,34 @@ const normalizeProp = (prop, context) => {
...
+  if (x === y) {
+    return x !== 0 || 1 / x === 1 / y;
+  }
+  return x !== x && y !== y; // eslint-disable-line no-self-compare
+};
+
+const equals = (x, y) =>
+  typeof Object.is === 'function' ? Object.is(x, y) : objectIs(x, y);
+
+const useStateWithDeps = (initialStateFn, deps) => {
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/DataTable.js:67]: This approach feels a bit too generic-ish. I'm wondering instead if we should create a more specialized hook called something like `useGroupState` which pulls `buildGroupState` inside it and then has an effect that updates the state when `groups` or `groupBy` change.

Comment (watiko) [reply src/js/components/DataTable/DataTable.js]: Thank you for your review. I made `useGroupState` hook. Would you review this?

**Review by ericsoderberghp — APPROVED** [review]

**Comment by ericsoderberghp** [comment]
Thanks for your contribution!

**Pull request merged**